### DISCUSSION TITLE
Passing in a scalar Param into a generator

### DISCUSSION THREAD

**Discussion by icebeing** [halide/Halide#7998] [Q&A] [answered]
Hi all -

   In the algorithm I'm working on, I realized I need to pass in a scalar value into my generator() method.  Similar in concept to the "offset" Param that's in lesson_10*.   Also, this is not an ImageParam, nor a GeneratorParam, if there's confusion in my meaning.

   However, my generator library is generated with the steps outlined in lesson_21 (where I don't invoke Func.compile_to_static_library(...) )

   So, my runner code knows the nature of the scalar, but how do I pass it into my generator prototype fct?

Thanks, Charles.

**Comment by abadams** [comment]
If I understand correctly, yyou're asking is for the Generator equivalent of a `Param<T>`. For generators, buffer inputs (ImageParams) are `Input<Buffer<some_type>>` and scalar inputs are `Input<some_type>`. E.g. in lesson 21, there's the floating point param `Input<float> factor{"factor"}`

If the value of the scalar is known at compile-time and you want to make compile-time decisions based on it, that's a GeneratorParam.

**Reply by icebeing** [reply, marked as answer]
Correct.  I was just looking at that lesson21 generator.  I guess I can interpret it this way: 

When you declare your Generator class, you have this sequence of public-scoped member variables:

```
class MyGenerator : public Halide::Generator<MyGenerator> 
{
    public:
    Input<Buffer<uint8_t>> input1;
    Input<Buffer<uint8_t>> input2;
    Input<int> someIntScalar;
    Output<Buffer<uint8_t>> output;

    ///etc...
};
```
When I run the GenGen.cpp - oriented mygenerator_generator executable, it effectively outputs this prototype in the mygen.h header file

```
HALIDE_FUNCTION_ATTRS
int mygenerator(struct halide_buffer_t *_input1, struct halide_buffer_t *_input2, int someIntScalar, struct halide_buffer_t *_output);

```
basically, you read these public member variables in the order they're declared in.  Also, I think why these are declared with this particular scoping too...

On another topic, I tried building the lesson21 generator, to see this header.  I got this error:

```
LD_LIBRARY_PATH=~/src/halide-install/lib ./lesson_21_generate -o . -g auto_schedule_gen -f auto_schedule_false -e static_library,h,schedule target=host auto_schedule=false 
Unhandled exception: Error: auto_schedule=true is no longer supported for enabling autoscheduling; specify autoscheduler=NAME instead.

```

Seems like the GenGen command lines need updating in the tutorials...

In any case, I think I understand what's happening.
